runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/SimpleParDoFn.java

```diff
@@ -486,14 +487,23 @@ private void processTimers(
     for (W window : windowsToCleanup) {
       // The stepContext is the thing that know if it is batch or streaming, hence
       // whether state needs to be cleaned up or will simply be discarded so the
-      // timer can be ignored
-
+      // timer can be ignored.
       Instant cleanupTime = earliestAllowableCleanupTime(window, windowingStrategy);
-      // if DoFn has OnWindowExpiration then set holds for system timer.
-      Instant cleanupOutputTimestamp =
-          fnSignature.onWindowExpiration() == null ? cleanupTime : cleanupTime.minus(1L);
-      stepContext.setStateCleanupTimer(
-          CLEANUP_TIMER_ID, window, windowCoder, cleanupTime, cleanupOutputTimestamp);
+      // Set a cleanup timer for state at the end of the window to trigger onWindowExpiration and
+      // garbage collect state. We avoid doing this for the global window if there is no window
+      // expiration set as the state will be up when the pipeline terminates. Setting the timer
+      // leads to a unbounded growth of timers for pipelines with many unique keys in the global
+      // window.
+      if (cleanupTime.isBefore(GlobalWindow.INSTANCE.maxTimestamp())
```

Comment:
Re-reading the bug and thinking about your comments, can we eliminate this condition and invoke from some `onDrain` method?

Reply:
I'm not sure I understand your suggestion. 
If we eliminate this condition, we are back to always setting timers which ends up accumulating timers for the global window.  If we eliminate setting the timer if the user did not specify onWindowExpiration, we are removing the automatic cleanup of windowed user-state for the non global window.

Are you are concerned about setting the timers if onWindowExpiration was requested for the global window? That is true, but at least the user can work around that by removing the onWindowExpiration annotation and modifying their code. Currently there is no mechanism in the dataflow streaming backend to scan and delivering an onDrain notification for all keys with user state, though that would be possible to add.  Any onDrain mechanism currently would have to be powered by timers set for the end of time, which would have the same issues as the cleanup timer.

Reply:
I meant the latter - concerned about setting timers if `onWindowExpiration` was requested for the global window.

I meant to imply us implementing some internal `onDrain` callback just for `ParDoFn` to receive. The backend would have to store all the keys & windows that need a callback, indeed. Perhaps just loudly warning that they have submitted a pipeline that could be trouble is best.

At some point we probably do want a user-facing `onDrain` rather than having them inspect the `PaneInfo` but that is a bigger conversation than this change.